src/interface/mod.rs

```diff
@@ -20,368 +20,608 @@
+        self.version_minor
+    }
 
-#[derive(Debug)]
-pub(crate) enum Mapping<'a> {
-    Datastream(&'a DatastreamMapping),
-    Properties(&'a PropertiesMapping),
-}
+    /// Returns the interface version.
+    fn version(&self) -> (i32, i32) {
+        (self.version_major, self.version_minor)
+    }
 
-#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
-pub(crate) struct BaseMapping {
-    endpoint: String,
-    #[serde(rename = "type")]
-    mapping_type: MappingType,
-    #[serde(skip_serializing_if = "Option::is_none")]
-    description: Option<String>,
-    #[serde(skip_serializing_if = "Option::is_none")]
-    doc: Option<String>,
-}
+    /// Returns the interface type.
+    pub fn interface_type(&self) -> InterfaceType {
+        match &self.inner {
+            InterfaceIml::DatastreamIndividual(_) => InterfaceType::Datastream,
+            InterfaceIml::DatastreamObject(_) => InterfaceType::Datastream,
+            InterfaceIml::Properties(_) => InterfaceType::Properties,
+        }
+    }
+
+    /// Returns the interface ownership.
+    pub fn ownership(&self) -> Ownership {
+        self.ownership
+    }
+
+    /// Returns the interface aggregation.
+    pub fn aggregation(&self) -> Aggregation {
+        match &self.inner {
+            InterfaceIml::DatastreamIndividual(_) => Aggregation::Individual,
+            InterfaceIml::DatastreamObject(_) => Aggregation::Object,
+            InterfaceIml::Properties(_) => Aggregation::Individual,
+        }
+    }
+
+    pub fn description(&self) -> Option<&str> {
+        self.description.as_deref()
+    }
+
+    pub fn doc(&self) -> Option<&str> {
+        self.doc.as_deref()
+    }
+
+    pub(crate) fn iter_mappings(&self) -> MappingIter {
+        MappingIter::new(&self.inner)
+    }
+
+    pub(crate) fn properties(&self) -> Option<&Properties> {
+        match &self.inner {
+            InterfaceIml::Properties(properties) => Some(properties),
+            _ => None,
+        }
+    }
+
+    pub(crate) fn mapping<'a: 's, 's>(&'s self, path: &MappingPath<'a>) -> Option<Mapping<'s>> {
+        match &self.inner {
+            InterfaceIml::DatastreamIndividual(individual) => individual.mapping(path),
+            InterfaceIml::DatastreamObject(object) => object.mapping(path),
+            InterfaceIml::Properties(properties) => properties.mapping(path),
+        }
+    }
+
+    pub(crate) fn contains(&self, path: &MappingPath<'_>) -> bool {
+        match &self.inner {
+            InterfaceIml::DatastreamIndividual(individual) => individual.contains(path),
+            InterfaceIml::DatastreamObject(object) => object.contains(path),
+            InterfaceIml::Properties(properties) => properties.contains(path),
+        }
+    }
+
+    pub fn mappings_len(&self) -> usize {
+        match &self.inner {
+            InterfaceIml::DatastreamIndividual(datastream) => datastream.mappings.len(),
+            InterfaceIml::DatastreamObject(datastream) => datastream.mappings.len(),
```

Comment:
```suggestion
            InterfaceIml::DatastreamIndividual(datastream) |
            InterfaceIml::DatastreamObject(datastream) => datastream.mappings.len(),
```

Reply:
The type on each side it's different, so in this case it's not possible to OR the matches